### DISCUSSION TITLE
Brownie Questions

### DISCUSSION THREAD

**Discussion by TyllerDaniel** [smartcontractkit/full-blockchain-solidity-course-py#139] [Q&A] [answered]
I have followed all the steps in installing Brownie and i am getting this 
👇👇👇
PS C:\Users\Gachengoh\demos\brownie_simple_storage> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.16.4 - Python development framework for Ethereum

When i try compiling the code, I get this errors
👇👇👇👇👇
File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\__main__.py", line 64, in main
    importlib.import_module(f"brownie._cli.{cmd}").main()
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\compile.py", line 50, in main
    proj = project.load()
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\main.py", line 750, in load
    return Project(name, project_path)
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\main.py", line 182, in __init__
    self.load()
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\main.py", line 237, in load
    self._compile(changed, self._compiler_config, False)
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\main.py", line 94, in _compile
    build_json = compiler.compile_and_format(
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\compiler\__init__.py", line 105, in compile_and_format
    find_solc_versions(solc_sources, install_needed=True, silent=silent)
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\compiler\solidity.py", line 150, in find_solc_versions
    pragma_specs[path] = sources.get_pragma_spec(source, path)
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\sources.py", line 218, in get_pragma_spec
    return NpmSpec(pragma_string)
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\semantic_version\base.py", line 618, in __init__
    self.clause = self._parse_to_clause(expression)
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\semantic_version\base.py", line 1225, in _parse_to_clause
    return cls.Parser.parse(expression)
  File "C:\Users\Gachengoh\.local\pipx\venvs\eth-brownie\lib\site-packages\semantic_version\base.py", line 1265, in parse
    raise ValueError("Invalid NPM block in %r: %r" % (expression, block))
ValueError: Invalid NPM block in '>= 0.6.0 < 0.9.0': '>='

Any help would be appreciated

**Comment by cromewar** [comment, marked as answer]
ValueError: Invalid NPM block in '>= 0.6.0 < 0.9.0': '>=' it seems to be a problem related to the compiler:

1. try to properly select the compiler version from vscode extension
2. make sure you don't have empty spaces on the sol version declaration, copy and paste this: (pragma solidity >=0.6.6 <0.9.0;)
3. If nothing works, uninstall brownie from pipx and install it on a python virtual environment using pip.

**Comment by TyllerDaniel** [comment]
Bro you are a life saver.Thanks a lot. I uninstalled pipx and installed the brownie  using pip

**Reply by cromewar** [reply]
You're welcome, I'm glad it helped you.

**Reply by PatrickAlphaC** [reply]
Boom! Great work both of you!

**Comment by leopashov** [comment]
I think I'm having a similar issue but wasn't able to fix it with cromewar's suggestion above. 
The code is spitting out: 
`(eth) C:\Users\leopa\Documents\Python Scripts\Patrick Collins\smartcontract-lottery>brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.2 - Python development framework for Ethereum

  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\_cli\__main__.py", line 64, in main
    importlib.import_module(f"brownie._cli.{cmd}").main()
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\_cli\compile.py", line 50, in main
    proj = project.load()
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 751, in load
    return Project(name, project_path)
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 183, in __init__
    self.load()
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 238, in load
    self._compile(changed, self._compiler_config, False)
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 90, in _compile
    _install_dependencies(self._path)
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 757, in _install_dependencies
    install_package(package_id)
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 779, in install_package
    return _install_from_github(package_id)
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 841, in _install_from_github
    download_url = _get_download_url_from_tag(org, repo, version, headers)
  File "C:\Users\leopa\anaconda3\envs\eth\lib\site-packages\brownie\project\main.py", line 912, in _get_download_url_from_tag
    raise ValueError(
ValueError: Invalid version for this package. Available versions are:
0.2.1, test0.2.1, 1.2.0, 1.1.1, 1.1.0, 1.0.2, 1.0.1, 1.0.0, 0.2.2, 0.2.1, 0.2.0`

Any more ideas with how to fix this?

**Reply by cromewar** [reply]
Hello @leopashov I see you are using anaconda envs, which are great for ML and Data Science, but I do not recommend them for Solidity dev, try using simple python virtual envs ;)

**Reply by leopashov** [reply]
Thank you for getting back to me. I created a new python venv within my smart contract lottery folder and installed brownie to it using pip but am getting the same error.
Was my method of creating the python venv incorrect or could there be another issue?

**Reply by leopashov** [reply]
I am now getting one problem from VSCode:
Unused function parameter. Remove or comment out the variable name to silence this warning.
from line:
function fulfillRandomness(bytes32 _requestId, uint256 _randomness)
relating to the _requestId function parameter.

I have cross referenced my code with the github repo and have found no differences

I am still getting the 'ValueError: Invalid version for this package. Available versions are:
0.2.1, test0.2.1, 1.2.0, 1.1.1, 1.1.0, 1.0.2, 1.0.1, 1.0.0, 0.2.2, 0.2.1, 0.2.0' error in terminal

**Reply by cromewar** [reply]
Hello @leopashov it seems a python library is outdated on windows so is giving troubles with brownie, you could try to upgrade the pip packages using [this](https://www.activestate.com/resources/quick-reads/how-to-update-all-python-packages/) guide, or try to install WSL and manage your python libraries there.

**Reply by leopashov** [reply]
Hi, I upgraded all packages but no improvement, I uninstalled anaconda and python from my pc, reinstalled python reinstalled web3 and brownie using pip and am still getting the same. I am extremely confused lol

**Reply by cromewar** [reply]
Windows is mess at python packages manipulation, did you consider to use WSL? you can find a guide about it here: https://github.com/smartcontractkit/full-blockchain-solidity-course-py#windows-support

**Reply by leopashov** [reply]
Ooh Ok ill definitely give that a go, thank you! Actually quite glad to hear it's a windows issue, as annoying as it is

**Reply by cromewar** [reply]
Yep that's why I stopped to use windows for development. I now I use almost exclusively WSL.

**Reply by leopashov** [reply]
@cromewar I ran through your tutorial but I'm not sure what's next? Do I use vscode as I was previously (Ie typing in the vscode command line) or should I be using WSL? I feel I've downloaded and installed all the stuff but don't know how to use it. 
Also should I be adding the brownie path into my windows PATH?
Many thanks

**Reply by leopashov** [reply]
am i right to just be switching 'cmd' to 'wsl' in the vscode terminal and going from there, for example?

I tried running brownie compile in vs code using wsl and got:
FileNotFoundError: [Errno 2] No such file or directory: 'C:/Users/leopa/.brownie/packages/OpenZeppelin/openzeppelin-contracts@3.4.0/contracts/access/Ownable.sol'
which looks less bad than the original error, however I don't know why it's searching this path. 
my import command in the sol file is: `import "@openzeppelin/contracts/access/Ownable.sol";`
and I have:
`- '@openzeppelin = OpenZeppelin/openzeppelin-contracts@3.4.0'
under remappings in the config file

**Reply by cromewar** [reply]
Yes you are right, you should use vscode but using WSL if you look to the left bottom you'll see this button:
![image](https://user-images.githubusercontent.com/20199180/145259774-092973ca-5777-47dd-af6d-5863ec9f9824.png)

click it and run a new WSL windows, so vscode will be attached to it, and also the terminal will be the linux one.

For now on you can forget CMD for ever (not quite dramatic) and use WSL for managing the python stuff, additionally you can run vscode from the linux terminal typing  `code .`

**Reply by leopashov** [reply]
@cromewar I've started using WSL now and so far so good! 
Tried running my code again and got the same issue as above but then cloned patrick's code from Github and that compiled fine so I must have made a mistake somewhere in copying the code. Thank you very much for your help! I feel more commfortable now knowing things are installed correctly (hopefully!) etc.

**Reply by cromewar** [reply]
I'm glad the info was helpful for you. if you never used linux before I know it could be challenging to get started with WSL so if you face any other problems during the course feel free to make any questions on this repo.

**Comment by Teqshark** [comment]
Hi everyone I had this same issue and 

> pipx

 was not the culprit!
it was being caused by the comment i had at the top of the contract before declaring pragma: 

> // SPDX-License-Identifier: MIT
> // pragma solidity >=0.6.6 < 0.9.0;
> pragma solidity >=0.8.0;
> 
> contract SimpleStorage {
>     uint256 favoriteNumber;
> 
>  .... rest of contract 
> 
as you can see that comment was another pragma declaration.
after removing the comment in second line `// pragma solidity >=0.6.6 < 0.9.0;`
and ending with:

```
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.0;

contract SimpleStorage {
    uint256 favoriteNumber;

 .... rest of contract 

```

brownie compile worked like a charm!

it seems that brownie tried to use the comment when getting the pragma specs from the file and thus the compilation failed. 

I am relatively inexperienced with how to best proceed to report this bug to the team and it could cause unnecessary problems for other developers. Please advise..

**Reply by Teqshark** [reply]
I fiddled a little with the comment and after the removing the space after the `<` in 
```// pragma solidity >=0.6.6 < 0.9.0;```
and ending with 
```//pragma solidity >=0.6.6 <0.9.0;```
it compiled successfully as well.

**Reply by Teqshark** [reply]
for some reason the syntax error in the original comment made brownie not compile the contract which is strange I guess brownie is making sure the pragma call in the comment is syntactically correct.

**Reply by Teqshark** [reply]
again from the logs it seems brownie is using the commented pragma declaration in the `get_pragma_spec` call in `brownie/project/sources.py` when compiling 
so since that was syntactically incorrect because it had a space after the `>` it threw the 
```
raise ValueError("Invalid NPM block in %r: %r" % (expression, block))
ValueError: Invalid NPM block in '>=0.6.6 < 0.9.0': '<' 
```
still i don't think brownie should be checking the commented pragma declaration at all.

**Reply by mth0801** [reply]
Thanks for your clarification, I had the same problem. After reading your answer, the error problem was solved.
